google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,469 @@
+   * Write one request to the streaming RPC.
+   *
+   * @note Only **one** `Write()` operation may be pending at a time. The
+   * application is responsible for waiting until any previous `Write()`
+   * operations have completed before calling `Write()` again.
+   *
+   * Whether `Write()` can be called before waiting for a matching `Read()`
+   * operation is specified by each service and RPC. Many services tolerate
+   * multiple `Write()` calls before performing or at least receiving a `Read()`
+   * response.
+   *
+   * If `true` is returned, the call was successful. If `false` is returned,
+   * the call failed, but the user may call `Write` again unless `Start` had
+   * finished with a permanent error `Status`. The application should wait until
+   * any other pending operations (typically any other `Read()` calls) complete
+   * and then call `Finish()` to find the status of the streaming RPC.
+   */
+  virtual future<bool> Write(Request const&, grpc::WriteOptions) = 0;
+
+  /**
+   * Return the final status of the streaming RPC.
+   *
+   * The application must wait until all pending `Read()` and `Write()`
+   * operations have completed before calling `Finish()`.
+   */
+  virtual future<Status> Finish() = 0;
+};
+
+template <typename RequestType, typename ResponseType>
+class ResumableAsyncStreamingReadWriteRpcImpl
+    : public ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      std::shared_ptr<RetryPolicy const> retry_policy,
+      std::shared_ptr<BackoffPolicy const> backoff_policy,
+      AsyncSleeper const sleeper,
+      AsyncStreamFactory<RequestType, ResponseType> stream_factory,
+      StreamInitializer<RequestType, ResponseType> initializer)
+      : retry_policy_prototype_(std::move(retry_policy)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        initializer_(std::move(initializer)) {}
+
+  ResumableAsyncStreamingReadWriteRpcImpl(ResumableStreamingReadRpc&&) = delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableStreamingReadRpc&&) = delete;
+
+  future<Status> Start() override {
+    future<Status> status_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      status_future = status_promise_.get_future();
+    }
+    Initialize(retry_policy_prototype_->clone(),
+               backoff_policy_prototype_->clone());
+    return status_future;
+  }
+
+  future<absl::optional<ResponseType>> Read() override {
+    future<absl::optional<ResponseType>> read_future;
+    bool retrying = false;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      retrying = stream_state_ == State::kRetrying;
+      if (retrying) {
+        read_reinit_done_.emplace(absl::make_optional(promise<void>()));
+        read_future = read_reinit_done_->get_future();
+      } else {
+        read_future = stream_->Read();
```

Comment:
If you're shutdown, this might be nullptr. Instead, make this a switch on the 3 enum states. shutdown: bail out with nullopt, retrying: what you do here then bail out (return read_future inline)